Add Vault migration post (#26)
bespinian/blog: content/posts/vault-migration/index.md

---
title: "Migrating HashiCorp Vault Between AKS Clusters"
date: 2025-03-18
author: "Max Leske (Xovis) and Mathis Kretz (bespinian)"
tags: ["Kubernetes", "Vault", "Azure", "Cloud Engineering"]
categories: ["Engineering", "Cloud Native"]
---

# Migrating HashiCorp Vault Between AKS Clusters

## Introduction

At bespinian and Xovis, we recently faced the challenge of migrating a
Kubernetes-based HashiCorp Vault instance from one Azure Kubernetes Service
(AKS) cluster to another. The goal was to consolidate infrastructure and reduce
the number of AKS clusters that required maintenance. To achieve this seamlessly
and reliably, we implemented the entire migration process as an automated Bash

> script.

Reviewer: ```suggestion
the number of AKS clusters that required maintenance. The outcome of a somewhat arduous journey is a Bash script that has enabled us to test the migration repeatedly and then perform the migration with a outage of less than 2 minutes.
```

By leveraging Vault’s Raft storage and high availability (HA) capabilities, we
scripted the migration end-to-end — from snapshot creation to restoring the
state, migrating unseal keys, and forwarding DNS traffic to the new instance.
This automation allowed us to minimize human error, enforce a repeatable

> process, and ensure minimal downtime.

Reviewer: ```suggestion
```

This blog post walks through our approach and presents the migration steps in 14
structured chapters. Each chapter includes excerpts from the automation script,
highlighting the checks, operations, and configuration changes we performed
along the way.

## Why We Migrated

Operating multiple AKS clusters can introduce unnecessary complexity and
overhead. To optimize our infrastructure and simplify maintenance, we decided to
migrate Vault from one AKS cluster to another, reducing the number of clusters
we needed to manage. Our primary goals were:

- Minimal downtime to avoid disruptions to dependent services.
- Data integrity with a verifiable transition from the old to the new instance.

> - Seamless DNS switchover to ensure applications continued to function without

Reviewer: ```suggestion
- Seamless switchover to ensure applications continued to function without
```

  configuration changes.
- A rollback strategy in case of unforeseen issues.

## Migration Strategy

We leveraged Vault’s built-in **Raft snapshot and restore** mechanism, which
allowed us to take a snapshot of the existing Vault data and restore it into a
new instance on the target AKS cluster. The migration was automated using a Bash
script that handled key steps such as Vault snapshot creation, restoration, and

> unseal key migration.

Reviewer: I think here we should rather give an overview of the setup. I also think we should say that we're running Vault in HA with a Raft backend and auto-unseal with Azure KeyVault, rather than saying that we used Raft for the migration.

Author: Agreed. Changed this paragraph.

### Pre-Migration Setup

Before initiating the migration:

1. **Provision the new Vault cluster**: Set up a fresh Vault instance on the
   target AKS cluster with the same configuration as the existing one.
2. **Configure networking and authentication**: Ensure that the new instance has
   the necessary access permissions and that external clients can connect

>    post-migration.

Reviewer: ```suggestion
   the necessary access permissions and that external clients will be able to connect
   post-migration.
```

3. **Test the migration in a staging environment**: Migrate a test Vault

>    instance within a lower environment to verify the process.

Reviewer: ```suggestion
   instance within between two non-production clusters to verify the process.
```

environment and ensure all prerequisites are met. This step includes setting up
local backups, verifying Kubernetes contexts, ensuring proper access
permissions, and preparing the user for the maintenance window.

- Create a local backup directory:

  ```bash
  mkdir -p vault-backups
  ```

- Check that the `kubectl` contexts for both clusters are available

  ```bash
  if ! (kubectl config get-contexts "${OLD_K8S_NAME}" > /dev/null 2>&1 && \
      kubectl config get-contexts "${NEW_K8S_NAME}" > /dev/null 2>&1); then
    print_message "No contexts for ${OLD_K8S_NAME} and ${NEW_K8S_NAME}."
    print_message "Make sure you have credentials for both clusters"
    exit 1
  fi
  ```

- Check that both `kubectl` contexts are accessible

  ```bash
  switch_context "${OLD_K8S_NAME}" "${OLD_VAULT_NAME}"
  if ! kubectl get cm > /dev/null 2>&1; then
    print_message "Can't access resources on ${OLD_K8S_NAME}. Aborting"
    exit 1
  fi

  switch_context "${NEW_K8S_NAME}" "${NEW_VAULT_NAME}"
  if ! kubectl get cm > /dev/null 2>&1; then
    print_message "Can't access resources on ${NEW_K8S_NAME}. Aborting"
    exit 1
  fi
  ```

> 

Reviewer: We should add a step that checks for required tools: az, kubectl, `kubectl wait`, jq, ed, dig.

Author: Good idea. Added.

- Check that the new Vault cluster has three replicas

  ```bash
  availableReplicas=$(kubectl get statefulsets.apps "${NEW_VAULT_NAME}" \
    -o template --template="{{.status.availableReplicas}}")
  if [ "${availableReplicas}" -ge 3 ]; then
    print_message "Expected number of replicas found on ${NEW_K8S_NAME}"
  else
    print_message "Unexpected replica count on ${NEW_K8S_NAME}. Exiting."
    exit 1
  fi

>   ```

Reviewer: This is still in the script but technically no longer necessary. I think we should clean that up for the blog post and simply iterate over pods selected by labels.

Author: I would leave it like this. Using the status of the StatefulSet is preferable from my point of view, since it uses the same information as Kubernetes itself uses.

Reviewer: That's also ok, I just mean that we don't need to restrict the number of replicas to 3.

Author: Not sure if I understand correctly, can you make a suggestion @theseion ?

- Check that the Managed Service Identity of the new Vault cluster has access to

>   the unseal keys of both Vault instances

Reviewer: Here we should elaborate which role assignments are necessary. I can add that.

Author: Cool, yes please @theseion

>   the unseal keys of both Vault instances

Reviewer: Mention that this is part of the Azure auto-unseal implementation.

Author: Agreed. Changed this paragraph.

      --query "[?principalId == '${msi_principal_id}']" > /dev/null 2>&1; then
    print_message "Role assigment missing for new vault MSI"
    print_message "Assign role 'Key Vault Crypto Service Encryption User'"
    exit 1
  fi

  az account set -s "${OLD_VAULT_SUBSCRIPTION_NAME}"

  key_vault_id=$(az keyvault show --name "${OLD_KEY_VAULT_NAME}" \
    --query "id" | jq -r '.')
  if
      ! az role assignment list \
      --scope "${key_vault_id}" \
      --role "Key Vault Crypto Service Encryption User" \
      --query "[?principalId == '${msi_principal_id}']" > /dev/null 2>&1; then
    print_message "Role assigment missing for new vault MSI"
    print_message "Assign role 'Key Vault Crypto Service Encryption User'"
    exit 1
  fi
  ```

- Remind the user to now announce the start of the maintenance window

  ```bash
  print_message "Please announce the start of the maintenance window."
  wait_for_any_key
  ```

### 2. Create temporary migration tokens

To perform the snapshot and restore operations securely, we create short-lived
Vault tokens with limited permissions. These tokens allow access to specific
endpoints like snapshot creation, restoration, sealing, and leadership

> operations.

Reviewer: ```suggestion
operations, which aren't granted to the existing policies.
```

- Write policies for authorizing the migration operations

  ```hcl
  # Policy allowing to step down Vault leader
  path "sys/step-down" {
   capabilities = ["update", "sudo"]
  }
  # Policy allowing to save snapshots
  path "sys/storage/raft/snapshot" {
   capabilities = [ "create", "read", "update", "list" ]
  }
  # Policy allowing to restore vault's snapshots
  path "sys/storage/raft/snapshot-force" {
   capabilities = [ "create", "read", "update", "list" ]
  }
  # Policy allowing to seal a vault
  path "sys/seal" {
   capabilities = ["update", "sudo"]

>   }

Reviewer: This was part of an experiment. We realized that we needed to block all outbound traffic from Vault to prevent any active credentials from being revoked during the transition phase. I settled on a different solution (patching the statefulset).
```suggestion
```

  kubectl delete ing vault --wait=true
  ```

### 4. Create snapshot of the old Vault instance

We now take a snapshot of the Vault data using the Raft backend's built-in
snapshot mechanism. After the snapshot is saved and downloaded, we immediately
disable the old Vault to avoid unintended behavior.

- Log in with the migration token

  ```bash
  kubectl exec -it "${OLD_VAULT_NAME}-${leader_index}" -- \
    vault login "${OLD_VAULT_ADMIN_TOKEN}"
  ```

- Create the snapshot in the leader pod

  ```bash
  kubectl exec -it "${OLD_VAULT_NAME}-${leader_index}" -- \
    vault operator raft snapshot save "${snapshot_filepath}"
  ```

- Dowload the snapshot

  ```bash
  kubectl cp "${OLD_VAULT_NAME}-${leader_index}":"${snapshot_filepath}" \
    "${BACKUPS_DIR}/${OLD_VAULT_SNAPSHOT_FILENAME}"

  kubectl exec -it "${OLD_VAULT_NAME}-${leader_index}" -- \
    rm "${snapshot_temp_filepath}"
  ```

- Disable the old Vault instance by changing the command of the Vault
  StatefulSet to `sleep`

  ```bash
  kubectl get statefulset vault -o yaml > \
  "${BACKUPS_DIR}/${OLD_VAULT_STATEFULSET_FILENAME}"
  cp "${BACKUPS_DIR}/${OLD_VAULT_STATEFULSET_FILENAME}" .

>   ed "${OLD_VAULT_STATEFULSET_FILENAME}" <<EOF

Reviewer: Here we should show before and after states of the modified file.

Author: @theseion it's probably easier for you to grab the yaml file from the actual environment. I no longer have access. Can you take care of this?

Author: I have done the same for the Vault ConfigMap in sections 6 and 7.

  /args:/
  +1,/-config=/c

>           - sleep 999999

Reviewer: Learned this recently :)
```suggestion
          - sleep infinity
```

  .
  w
  q
  EOF

  kubectl apply -f "${OLD_VAULT_STATEFULSET_FILENAME}"
  ```

- Restart all pods to pick up the patched configuration
  ```bash
  kubectl delete pod --selector=app.kubernetes.io/name=vault
  ```

### 5. Restore the snapshot in the new Vault instance

With the snapshot from the old instance in hand, we restore it to the new Vault
instance. This effectively brings the new Vault cluster to the same data state
as the old one.

> - Upload the snapshot the leader pod

Reviewer: ```suggestion
- Upload the snapshot the to leader pod
```

  ```bash
  kubectl cp "${BACKUPS_DIR}/${OLD_VAULT_SNAPSHOT_FILENAME}" \
    "vault-${leader_index}":"${snapshot_tmp}"
  ```

- Log in with the migration token

  ```bash
  kubectl exec -it "vault-${leader_index}" -- \
    vault login "${NEW_VAULT_MIGRATION_TOKEN}"
  ```

- Run the restore command
  ```bash
  kubectl exec -it "vault-${leader_index}" -- \
    sh -c "vault operator raft snapshot restore -force \"${snapshot_tmp}\""
  ```

### 6. Configure the new Vault cluster to use the old unseal key

> To allow the restored Vault data to function properly, we need to configure the
> new instance to use the same unseal key as the old Vault. This step modifies the
> config map and restarts the pods to apply the change.

Reviewer: ```suggestion
The restored state must be unsealed with the unseal key stored in the KeyVault associated with the _old_ Vault instance. We now need to unseal the state but store a new unseal key in the KeyVault associated with the _new_ Vault instance. This step modifies the config map by replacing the coordinates to the _new_ instance unseal key with those of the _old_ instance unseal key and restarts the pods to apply the change. After the restart, the pods should unseal the state automatically since they have access to the original unseal key.
```

- Patch the configuration of the new Vault cluster with the credentials of the
  old Azure Key Vault

  ```bash
  kubectl get cm vault-config -o yaml > "${vault_config_filename}"

  ed '+/seal "/' "${vault_config_filename}" <<EOF
  .,/}/d
  -
  a
      seal "azurekeyvault" {
        tenant_id      = "${TENANT_ID}"
        client_id      = "${NEW_VAULT_MSI_CLIENT_ID}"
        vault_name     = "${OLD_KEY_VAULT_NAME}"
        key_name       = "${OLD_VAULT_UNSEAL_KEY_SECRET_NAME}"
      }
  .
  w
  q
  EOF

  kubectl apply -f "${vault_config_filename}"
  ```

- Restart all pods to pick up the patched configuration
  ```bash
  kubectl delete pod --selector=app.kubernetes.io/name=vault
  ```

### 7. Prepare the unseal key migration on the new Vault instance

> To enable unseal key migration, we first need to configure the new Vault cluster
> to support dual seal configuration — keeping the old seal block active but
> disabled, and adding a new one that points to the new Azure Key Vault.

Reviewer: ```suggestion
To enable unseal key migration, the new Vault cluster needs to know about both unseal key coordinates when the configuration is loaded the next time (after unsealing the state) — keeping the old seal block active but
disabled, and adding a new one that points to the new Azure Key Vault.
```

- Update the Vault configuration:

```bash
kubectl get cm vault-config -o yaml > vault-config.yaml

ed '+/seal "/' vault-config.yaml <<EOF
a
      disabled       = true
.
/}/
a
    seal "azurekeyvault" {
      tenant_id      = "${TENANT_ID}"
      client_id      = "${NEW_VAULT_MSI_CLIENT_ID}"
      vault_name     = "${NEW_KEY_VAULT_NAME}"
      key_name       = "${NEW_VAULT_UNSEAL_KEY_SECRET_NAME}"
    }
.
w
q
EOF

kubectl apply -f vault-config.yaml
```

### 8. Perform the unseal key migration on the follower pods

Vault only allows unseal key migration on follower nodes, so we restart each
follower pod and run the migration steps.

> - Iterate through each follower pod and migrate using all three unseal keys:

Reviewer: Need to clarify the assumption of a 3 key threshold. We should also clarify "unseal key vs. recovery key".

Author: Done.

  ```bash

>   for i in {0..2}; do

Reviewer: Replace with pod selection by labels

Author: Can I leave that to you, @theseion? I no longer have a setup to test this effectively. Alternatively, we can just leave it as is for the blog post. It works and it is also quite easy to understand.

Reviewer: I'll take care of it. Goes hand in hand with the query to the StatefulSet at the top.

    if [ "${i}" -eq "${leader_index}" ]; then
      continue
    fi

    print_message "Restarting pod vault-${i} to start unseal key migration"
    kubectl delete pod "vault-${i}"
    kubectl wait \
      --for=jsonpath='{.status.containerStatuses[0].state.running}' \
      --selector=apps.kubernetes.io/pod-index="${i}" \
      --timeout=90s \
      pod

    print_message "Migrating vault-${i}"
    kubectl exec -it "vault-${i}" -- \
      sh -c "vault operator unseal -migrate \"${OLD_VAULT_UNSEAL_KEY_1}\""
    kubectl exec -it "vault-${i}" -- \
      sh -c "vault operator unseal -migrate \"${OLD_VAULT_UNSEAL_KEY_2}\""
    kubectl exec -it "vault-${i}" -- \
      sh -c "vault operator unseal -migrate \"${OLD_VAULT_UNSEAL_KEY_3}\""
  done
  ```

### 9. Step down leader pod

Once all follower pods have been migrated, the leader must step down so it can

> restart and apply the new seal configuration.

Reviewer: ```suggestion
restart and apply the new seal configuration.\
Note that we use the migration token of the _old_ Vault instance here, since that is the token stored in the restored snapshot.
```

- Step down the leader pod

  ```bash

>   kubectl exec -it vault-0 -- vault login "${OLD_VAULT_MIGRATION_TOKEN}"
>   kubectl exec -it vault-0 -- vault operator step-down

Reviewer: ```suggestion
  kubectl exec -it "vault-${leader-index}" -- vault login "${OLD_VAULT_MIGRATION_TOKEN}"
  kubectl exec -it "vault-${leader-index}" -- vault operator step-down
```

  ```

### 10. Remove old seal block from the config

Now that all pods use the new seal configuration, we can safely remove the old
seal block.

- Edit the config map

  ```bash
  kubectl get cm vault-config -o yaml > vault-config.yaml

  ed '+/disabled \+= true/' vault-config.yaml <<EOF
  -
  .;/}/d
  w
  q
  EOF

  kubectl apply -f vault-config.yaml
  ```

### 11. Restart all replicas

- Restart all Vault pods to ensure they all use the final configuration.
  ```bash
  kubectl delete pod --selector=app.kubernetes.io/name=vault
  ```

### 12. Create a proxy service from the old Vault to the new Vault

To allow clients using the old Vault domain to reach the new Vault instance, we
create a proxy `Service` and `EndpointSlice` in the old cluster that forwards

> traffic to the new Vault’s ingress.

Reviewer: We should probably mention that `ExternalName` in ingress-nginx is limited to licensed uses of nginx, hence we need to use an IP address instead of a DNS name.

Author: Agreed. Done.

```bash
new_vault_ip="$(dig +short "${NEW_VAULT_DOMAIN}" | tail -1)"

kubectl apply -f - <<EOF
---
apiVersion: v1
kind: Service
metadata:
  name: vault-migration-proxy
  namespace: vault
spec:
  ports:
    - port: 443
      protocol: TCP
      targetPort: 443
      name: https
---
apiVersion: discovery.k8s.io/v1
kind: EndpointSlice
metadata:
  name: vault-migration-proxy
  namespace: vault
  labels:
    kubernetes.io/service-name: vault-migration-proxy
addressType: IPv4
ports:
  - name: "https"
    port: 443
endpoints:
  - addresses:
      - "${new_vault_ip}"
EOF
```

### 13. Modify the ingress of the old Vault to forward connections via proxy

> With the proxy service in place, we modify the existing ingress to forward all
> requests to the proxy.

Reviewer: ```suggestion
With the proxy service in place, we recreate the ingress for the old Vault instance but configured to forward all
requests to the proxy.
```

- Edit the old ingress and update

  ```bash
  cp "${BACKUPS_DIR}/${OLD_VAULT_INGRESS_FILENAME}" "${new_ingress_filename}"
  ed '+/annotations:/' "${new_ingress_filename}" <<EOF
  a
      nginx.ingress.kubernetes.io/backend-protocol: HTTPS
      nginx.ingress.kubernetes.io/ssl-redirect: "true"
      nginx.ingress.kubernetes.io/upstream-vhost: "${NEW_VAULT_DOMAIN}"
  .
  /name: vault-active/
  s/name:.*/name: vault-migration-proxy
  /number: 8200/
  s/number:.*/number: 443
  w
  q
  EOF

  kubectl apply -f "${new_ingress_filename}"
  ```

>   This allows existing clients to continue using the old Vault domain without
>   any modifications.

Reviewer: ```suggestion
  This allows existing clients to continue using the old Vault domain without
  any modifications, as long as the configuration on the old cluster forwards requests.
```

### 14. Test the reachability of the old Vault domain

Finally, we test that both the old and new Vault domains point to the same

> instance and that authentication works as expected.

Reviewer: ```suggestion
instance and that authentication works as expected.\
Note again the use of the _old_ Vault token for both tests.
```
